### PR TITLE
feat: add auto-wired metrics middleware for GCP API calls

### PR DESCRIPTION
**Pull request by stekik**
**Description**
1. **gRPC Interceptor** (NEW pattern clients)
2. **HTTP RoundTripper** (OLD pattern REST clients)

Changes proposed in this pull request:
- **Clean business logic** - zero metric calls in feature code
- **Automatic metering** - all API calls tracked without manual intervention
- **Complete labels** - region/project auto-populated via reflection

**Related issue(s)**

https://github.com/kyma-project/cloud-manager/issues/1637

### CODE REVIEW

**Review by dushanpantic — COMMENTED** [review]
Have you tested if this actually works?
I am not sure if `WithGRPCDialOption` is compatible with `NewNetworksRESTClient`

File: pkg/kcp/provider/gcp/client/metrics.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	return region, project
+}
+
+// Example: /v1/projects/my-project/locations/europe-west1/instances/my-instance -> ("europe-west1", "my-project")
+func extractRegionAndProjectFromURL(path string) (region, project string) {
+	parts := strings.Split(strings.TrimPrefix(path, "/"), "/")
+
+	for i := 0; i < len(parts)-1; i++ {
```
[the remark this answers is not in the document]
Comment (stekik) [reply pkg/kcp/provider/gcp/client/metrics.go]: Changed loop to iterate over all parts

File: pkg/kcp/provider/gcp/client/metrics.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+			}
+		case "regions", "locations":
+			if i+1 < len(parts) && !strings.HasPrefix(parts[i+1], "{") {
+				region = parts[i+1]
+			}
+		case "zones":
+			if i+1 < len(parts) && !strings.HasPrefix(parts[i+1], "{") {
+				zone := parts[i+1]
+				if idx := strings.LastIndexByte(zone, '-'); idx > 0 {
+					region = zone[:idx]
```
[the remark this answers is not in the document]
Comment (stekik) [reply pkg/kcp/provider/gcp/client/metrics.go]: Added validation for content after hyphen

File: pkg/kcp/provider/gcp/client/metrics.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+			return false
+		}
+	}
+
+	return !hasHyphen
+}
+
+// hasResourceID checks if the path targets a specific resource
+func hasResourceID(parts []string) bool {
+	return len(parts) > 1 && !strings.HasPrefix(parts[len(parts)-1], "{")
```
[the remark this answers is not in the document]
Comment (stekik) [reply pkg/kcp/provider/gcp/client/metrics.go]: Rewrote logic to handle all custom method patterns

File: pkg/kcp/provider/gcp/client/metrics.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+// Example: POST /v1/services/{service}/connections:deleteConnection -> "Connections.DeleteConnection"
+func extractOperationFromURL(path, method string) string {
+	parts := filterPathSegments(strings.Split(strings.TrimPrefix(path, "/"), "/"))
+	if len(parts) == 0 {
+		return method
+	}
+
+	// Handle custom methods (":enable", ":disable", ":deleteConnection", etc.)
+	lastPart := parts[len(parts)-1]
+	if idx := strings.IndexByte(lastPart, ':'); idx > 0 {
```
[the remark this answers is not in the document]
Comment (stekik) [reply pkg/kcp/provider/gcp/client/metrics.go]: Boundary check fixed

**Review by tmilos77 — COMMENTED** [review]
It seems like a complication to have middlewares/interceptors/roundtrippers with a state having the service name. The `google.cloud.compute.v1.Addresses/Insert` is perfectly fine value for the metrics operation attribute. Is it possible to refactor IncrementCallCounter to just accept the method argument instead of serviceName, operationName, and have to extracted solely from the request, so we don't need to have different clients created with different serviceName?

### Example:

Instead of 
```
// one.go
httpClient := gcpclient.NewMetricsHTTPClient("Compute", baseClient.Transport)
// second.go
httpClient := client.NewMetricsHTTPClient("ServiceNetworking", baseClient.Transport)
// third.go
httpClient := client.NewMetricsHTTPClient("File", baseClient.Transport)
```

all using request extractor providing operation from the request, ie  `/google.cloud.compute.v1.Addresses/Insert` => `Insert` and adding resource name to it, and then using `IncrementCallCounter` that does 
`method = serviceName + operation`, ie 
* Compute.VM.Insert
* ServiceNetworking.Something.Delete
* File.FileStore.Insert

we actually do
```
// one.go
httpClient := NewMetricsClient(baseClient.Transport)
// second.go
httpClient := NewMetricsClient(baseClient.Transport)
// third.go
httpClient := NewMetricsClient(baseClient.Transport)
```

and then use extractor that simply uses request to get the operation, ie `google.cloud.compute.v1.Addresses/Insert` where that already has the service name, resource name, and the operation?

**Review by tmilos77 — CHANGES_REQUESTED** [review]
After a second thought, this is just to complicated, we should change this. Do not bother identifying parts of method (service, resource, operation), just use raw request data you have, like http method+path

File: pkg/kcp/provider/gcp/client/metrics.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	if err != nil {
+		return err
+	}
+	if resp != nil && resp.StatusCode >= 400 {
+		return &googleapi.Error{Code: resp.StatusCode}
+	}
+	return nil
+}
+
+func parseGoogRequestParams(params string) (region, project string) {
```
Comment (tmilos77) [inline pkg/kcp/provider/gcp/client/metrics.go]: it would be nice there's a test that provides example values and proves func works correctly in all cases

**Review by tmilos77 — CHANGES_REQUESTED** [review]
Much better now w/out reflection. Just provide some tests for those extraction functions that will provide example values for easier future maintenance and prove they work in edge cases

**Review by tmilos77 — APPROVED** [review]
👍

**Review by vanja-r — APPROVED** [review]
